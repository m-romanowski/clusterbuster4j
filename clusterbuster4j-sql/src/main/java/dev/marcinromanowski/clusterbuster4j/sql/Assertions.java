package dev.marcinromanowski.clusterbuster4j.sql;

import static java.util.Objects.isNull;

class Assertions {

    private Assertions() {

    }

    static void assertNonNull(Object value, String propertyName) {
        if (isNull(value)) {
            throw new IllegalArgumentException("%s must be not null".formatted(propertyName));
        }
    }

    static void assertGreaterEqualTo(int expected, int actual, String propertyName) {
        if (expected > actual) {
            throw new IllegalArgumentException("%s must be greater than or equal to %s".formatted(propertyName, expected));
        }
    }

}
