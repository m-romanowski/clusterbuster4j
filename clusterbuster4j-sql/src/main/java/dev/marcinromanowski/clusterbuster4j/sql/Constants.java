package dev.marcinromanowski.clusterbuster4j.sql;

class Constants {

    static final String CENTER_COL = "center";
    static final String CLUSTER_COL = "cluster";
    static final String PREV_CLUSTER_NO_COL = "previous_cluster_no";
    static final String CLUSTER_SIZE_COL = "cluster_size";
    static final String CLUSTER_NO_COL = "cluster_no";
    static final String LNG_COL = "longitude";
    static final String LAT_COL = "latitude";

    private Constants() {

    }

}
