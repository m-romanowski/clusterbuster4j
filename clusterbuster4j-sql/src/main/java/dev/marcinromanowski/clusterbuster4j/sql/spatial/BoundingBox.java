package dev.marcinromanowski.clusterbuster4j.sql.spatial;

public record BoundingBox(LatLng ne, LatLng sw) {

    public LatLng center() {
        final double latNorth = ne.lat();
        final double latSouth = sw.lat();
        final double lngEast = ne.lng();
        final double lngWest = sw.lng();
        return new LatLng((latNorth + latSouth) / 2, (lngEast + lngWest) / 2);
    }

}
