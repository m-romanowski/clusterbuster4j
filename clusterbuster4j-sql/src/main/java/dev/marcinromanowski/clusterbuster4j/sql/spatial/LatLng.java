package dev.marcinromanowski.clusterbuster4j.sql.spatial;

public record LatLng(double lat, double lng) {

    private static final int EARTH_RADIUS_KM = 6_371;
    private static final int MIN_LATITUDE = -90;
    private static final int MAX_LATITUDE = 90;
    private static final int MIN_LONGITUDE = -180;
    private static final int MAX_LONGITUDE = 180;

    public LatLng {
        if (lat < MIN_LATITUDE || lat > MAX_LATITUDE) {
            throw new IllegalArgumentException("Latitude must be in range [-90, 90]");
        }

        if (lng < MIN_LONGITUDE || lng > MAX_LONGITUDE) {
            throw new IllegalArgumentException("Longitude must be in range [-180, 180]");
        }
    }

    private static double haversine(double x) {
        final double sinHalf = Math.sin(x * 0.5);
        return sinHalf * sinHalf;
    }

    private static double arcHaversine(double x) {
        return 2 * Math.asin(Math.sqrt(x));
    }

    private static double haversineDistance(double lat1, double lat2, double dLng) {
        return haversine(lat1 - lat2) + haversine(dLng) * Math.cos(lat1) * Math.cos(lat2);
    }

    private static double distanceRadians(double lat1, double lng1, double lat2, double lng2) {
        return arcHaversine(haversineDistance(lat1, lat2, lng1 - lng2));
    }

    private static double computeAngleBetween(LatLng from, LatLng to) {
        return distanceRadians(Math.toRadians(from.lat()), Math.toRadians(from.lng()), Math.toRadians(to.lat()), Math.toRadians(to.lng()));
    }

    public double computeDistanceKm(LatLng to) {
        return computeAngleBetween(this, to) * EARTH_RADIUS_KM;
    }

}
