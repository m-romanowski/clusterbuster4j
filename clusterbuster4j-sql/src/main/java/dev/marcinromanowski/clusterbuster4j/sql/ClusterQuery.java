package dev.marcinromanowski.clusterbuster4j.sql;

import com.querydsl.core.Tuple;
import com.querydsl.sql.ProjectableSQLQuery;

import java.util.function.Supplier;

@FunctionalInterface
public interface ClusterQuery<T extends Tuple> extends Supplier<ProjectableSQLQuery<T, ?>> {

}
