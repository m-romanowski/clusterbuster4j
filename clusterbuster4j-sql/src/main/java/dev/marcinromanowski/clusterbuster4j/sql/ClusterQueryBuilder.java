package dev.marcinromanowski.clusterbuster4j.sql;

import com.querydsl.core.Tuple;
import com.querydsl.core.types.Expression;
import com.querydsl.core.types.FactoryExpression;
import com.querydsl.core.types.dsl.Expressions;
import com.querydsl.sql.AbstractSQLQueryFactory;
import com.querydsl.sql.ProjectableSQLQuery;
import com.querydsl.sql.SQLCommonQuery;
import com.querydsl.sql.SQLQuery;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.BoundingBox;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.DBSCANFormula;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.Zoom;
import org.geolatte.geom.Geometry;

import static dev.marcinromanowski.clusterbuster4j.sql.Assertions.assertGreaterEqualTo;
import static dev.marcinromanowski.clusterbuster4j.sql.Assertions.assertNonNull;

/**
 * Allows you to create a query that groups geometries and returns
 * clusters (groups) of geometries that are close to each other.
 * The algorithm primarily uses the bounding box and the current
 * zoom. Queries use CTE (<a href="https://www.postgresql.org/docs/current/queries-with.html">Common Table Expressions</a>), so
 * make sure your SQL database supports this query format.
 */
public class ClusterQueryBuilder {

    private final AbstractSQLQueryFactory<? extends ProjectableSQLQuery<?, ?>> queryFactory;

    private int zoom;
    private int minZoom;
    private int maxZoom;
    private BoundingBox bbox;
    private ProjectableSQLQuery<?, ?> baseQuery;
    @SuppressWarnings("rawtypes")
    private Expression<Geometry> geometryExpression;
    private DBSCANFormula dbScanFormula = new DBSCANFormulaImpl();

    private ClusterQueryBuilder(AbstractSQLQueryFactory<? extends ProjectableSQLQuery<?, ?>> factory) {
        this.queryFactory = factory;
    }

    public static ClusterQueryBuilder builder(AbstractSQLQueryFactory<? extends ProjectableSQLQuery<?, ?>> factory) {
        return new ClusterQueryBuilder(factory);
    }

    public ClusterQueryBuilder withZoom(int zoom) {
        this.zoom = zoom;
        return this;
    }

    public ClusterQueryBuilder withMinZoom(int minZoom) {
        assertGreaterEqualTo(0, minZoom, "minZoom");
        this.minZoom = minZoom;
        return this;
    }

    public ClusterQueryBuilder withMaxZoom(int maxZoom) {
        assertGreaterEqualTo(1, maxZoom, "maxZoom");
        this.maxZoom = maxZoom;
        return this;
    }

    public ClusterQueryBuilder withBoundingBox(BoundingBox bbox) {
        this.bbox = bbox;
        return this;
    }

    public ClusterQueryBuilder withBaseQuery(SQLQuery<?> baseQuery) {
        this.baseQuery = baseQuery;
        return this;
    }

    @SuppressWarnings("rawtypes")
    public ClusterQueryBuilder withGeometryExpression(Expression<Geometry> geometryExpression) {
        this.geometryExpression = geometryExpression;
        return this;
    }

    public ClusterQueryBuilder withDBSCANFormula(DBSCANFormula formula) {
        this.dbScanFormula = formula;
        return this;
    }

    public <T extends Tuple> ClusterQuery<T> build() {
        this.zoom = Math.min(zoom, maxZoom);
        this.minZoom = Math.min(Math.max(zoom, Math.max(0, minZoom)), maxZoom);

        assertNonNull(bbox, "Bounding box");
        assertNonNull(bbox.sw(), "SW coordinates");
        assertNonNull(bbox.ne(), "NE coordinates");
        assertNonNull(geometryExpression, "Geometry expression (SRID 4326)");
        assertNonNull(baseQuery, "Cluster base query");
        assertNonNull(dbScanFormula, "DBSCAN epsilon formula");

        return new SqlClusterQuery<T>(buildQuery());
    }

    @SuppressWarnings("unchecked")
    private <T extends Tuple> ProjectableSQLQuery<T, ?> buildQuery() {
        final Expression<?> projection = baseQuery.getMetadata()
            .getProjection();

        if (!(projection instanceof FactoryExpression<?> exprFactory)) {
            throw new IllegalArgumentException("The base query should be a factory expression, such as JavaBean or Constructor projections");
        }

        if (!(exprFactory.getArgs().contains(geometryExpression))) {
            throw new IllegalArgumentException("The base query should contain the geometry expression");
        }

        final BoundingBoxQuery bboxQuery = new BoundingBoxQuery(queryFactory, baseQuery, geometryExpression, bbox);
        final OutputClusterQuery out = new OutputClusterQuery(queryFactory, exprFactory);

        String queryName = bboxQuery.name();
        SQLCommonQuery<?> queryChain = out.query().with(Expressions.path(String.class, queryName), bboxQuery.query());
        for (int i = maxZoom; i >= zoom; i--) {
            final DBSCANCluster dbScanCluster = createDbScanClusterQueryFor(i, bbox, exprFactory, queryName, queryChain);
            final GroupedClusterQuery groupedCluster = createGroupedClusterQueryFor(i, exprFactory, dbScanCluster.name(), dbScanCluster.query());
            queryChain = groupedCluster.query();
            queryName = groupedCluster.name();
        }

        return (ProjectableSQLQuery<T, ?>) queryChain.from(Expressions.path(String.class, queryName));
    }

    private DBSCANCluster createDbScanClusterQueryFor(
        int zoom,
        BoundingBox bbox,
        FactoryExpression<?> projection,
        String baseQueryTableName,
        SQLCommonQuery<?> baseQuery
    ) {
        return new DBSCANCluster(
            queryFactory,
            zoom,
            dbScanFormula.apply(bbox, Zoom.of(zoom))
                .value(),
            projection,
            baseQueryTableName,
            baseQuery
        );
    }

    private GroupedClusterQuery createGroupedClusterQueryFor(
        int zoom,
        FactoryExpression<?> projection,
        String baseQueryTableName,
        SQLCommonQuery<?> baseQuery
    ) {
        return new GroupedClusterQuery(
            queryFactory,
            zoom,
            projection,
            baseQueryTableName,
            baseQuery
        );
    }

}
