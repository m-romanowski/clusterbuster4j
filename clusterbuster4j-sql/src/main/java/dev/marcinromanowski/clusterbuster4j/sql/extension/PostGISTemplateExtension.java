package dev.marcinromanowski.clusterbuster4j.sql.extension;

import com.querydsl.sql.spatial.PostGISTemplates;

public class PostGISTemplateExtension extends PostGISTemplates {

    public PostGISTemplateExtension() {
        this('\\', false);
    }

    public PostGISTemplateExtension(boolean quote) {
        this('\\', quote);
    }

    public PostGISTemplateExtension(char escape, boolean quote) {
        super(escape, quote);
        this.add(SqlOpsExtension.FIRST, "FIRST({0})");
        this.add(SqlOpsExtension.DBSCAN, "ST_ClusterDBSCAN({0}, {1}, {2}) over ()");
        this.add(SqlOpsExtension.MAKE_ENVELOPE, "ST_MakeEnvelope({0}, {1}, {2}, {3}, {4})");
    }

}
