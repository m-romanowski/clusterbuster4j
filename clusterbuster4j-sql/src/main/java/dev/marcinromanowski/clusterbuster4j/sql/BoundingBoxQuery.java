package dev.marcinromanowski.clusterbuster4j.sql;

import com.querydsl.core.types.Expression;
import com.querydsl.core.types.dsl.Expressions;
import com.querydsl.spatial.GeometryExpressions;
import com.querydsl.sql.AbstractSQLQueryFactory;
import com.querydsl.sql.ProjectableSQLQuery;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.BoundingBox;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.LatLng;
import org.geolatte.geom.Geometry;

import static dev.marcinromanowski.clusterbuster4j.sql.Constants.CENTER_COL;
import static dev.marcinromanowski.clusterbuster4j.sql.Constants.CLUSTER_NO_COL;
import static dev.marcinromanowski.clusterbuster4j.sql.Constants.CLUSTER_SIZE_COL;

/**
 * Returns all geometries that fall (intersects) within the specified bounding
 * box - the area bounded by NE (North-East) and SW (South-West) coordinates.
 */
class BoundingBoxQuery implements NamedQueryProvider {

    private static final String QUERY_NAME = "bbox_query";

    private final ProjectableSQLQuery<?, ?> query;

    @SuppressWarnings("rawtypes")
    BoundingBoxQuery(
        AbstractSQLQueryFactory<? extends ProjectableSQLQuery<?, ?>> queryFactory,
        ProjectableSQLQuery<?, ?> baseQuery,
        Expression<Geometry> geomExpr,
        BoundingBox bbox
    ) {
        final LatLng bboxNe = bbox.ne();
        final LatLng bboxSw = bbox.sw();
        this.query = queryFactory.select(
                Expressions.as(geomExpr, CENTER_COL),
                Expressions.asNumber(1)
                    .as(CLUSTER_SIZE_COL),
                Expressions.asNumber(0)
                    .as(CLUSTER_NO_COL),
                baseQuery.getMetadata()
                    .getProjection()
            )
            .from(Expressions.as(baseQuery, "base_query"))
            .where(
                GeometryExpressions.asGeometry(geomExpr)
                    // bbox = left, bottom, right, top
                    // bbox = min lng, min lat, max lng, max lat
                    .intersects(QueryExtensions.makeEnvelope(bboxSw.lng(), bboxSw.lat(), bboxNe.lng(), bboxNe.lat(), 4326))
            );
    }

    @Override
    public String name() {
        return QUERY_NAME;
    }

    @Override
    public ProjectableSQLQuery<?, ?> query() {
        return query;
    }

}
