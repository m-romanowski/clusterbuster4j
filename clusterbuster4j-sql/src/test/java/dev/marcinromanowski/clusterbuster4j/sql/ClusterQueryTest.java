package dev.marcinromanowski.clusterbuster4j.sql;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.querydsl.core.types.Expression;
import com.querydsl.core.types.dsl.Expressions;
import com.querydsl.core.types.dsl.SimplePath;
import com.querydsl.sql.Configuration;
import com.querydsl.sql.SQLQuery;
import com.querydsl.sql.SQLQueryFactory;
import dev.marcinromanowski.clusterbuster4j.fixture.SqlSnapshotSerializer;
import dev.marcinromanowski.clusterbuster4j.sql.extension.PostGISTemplateExtension;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.BoundingBox;
import dev.marcinromanowski.clusterbuster4j.sql.spatial.LatLng;
import org.geolatte.geom.Geometry;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

@ExtendWith(SnapshotExtension.class)
class ClusterQueryTest {

    private static final BoundingBox BOUNDING_BOX = new BoundingBox(new LatLng(17.9, 47.7), new LatLng(26.36, 50.61));

    @SuppressWarnings("unused")
    private Expect expect;
    private SQLQueryFactory sqlQueryFactory;
    @SuppressWarnings("rawtypes")
    private SimplePath<Geometry> geometryExpr;

    @BeforeEach
    void setUp() {
        sqlQueryFactory = queryFactory();
        geometryExpr = Expressions.path(Geometry.class, "geom");
    }

    @ParameterizedTest
    @CsvSource(delimiter = '|', textBlock = """
        19 | 20 | 15 | Min, max and current zoom should be normalized to valid values
        18 | 15 | 16 | Current zoom should be normalized to min zoom
        10 | 5  | 20 | All values are valid
        """
    )
    @SuppressWarnings("java:S2699") // Snapshot assertion
    void shouldAllowToCreateClusteringQuery(int zoom, int minZoom, int maxZoom, String reason) {
        // given
        final var query = baseQuery(geometryExpr);

        // when
        final var clusterQuery = ClusterQueryBuilder.builder(sqlQueryFactory)
            .withBaseQuery(query)
            .withBoundingBox(BOUNDING_BOX)
            .withZoom(zoom)
            .withMinZoom(minZoom)
            .withMaxZoom(maxZoom)
            .withGeometryExpression(geometryExpr)
            .build();

        // expect
        expect.scenario(reason)
            .serializer(SqlSnapshotSerializer.class)
            .toMatchSnapshot(clusterQuery.get().toString());
    }

    @Test
    void shouldReturnAnErrorWhenWeDoNotInitializeBuilderWithRequiredData() {
        // when
        final var builder = ClusterQueryBuilder.builder(sqlQueryFactory);

        // then
        assertThrows(IllegalArgumentException.class, builder::build);
    }

    @Test
    void shouldReturnAnErrorWhenWeDoNotSpecifyGeometryExpression() {
        // given
        final var query = baseQuery(geometryExpr);
        final var builder = ClusterQueryBuilder.builder(sqlQueryFactory)
            .withBaseQuery(query)
            .withBoundingBox(BOUNDING_BOX)
            .withZoom(10)
            .withMinZoom(5)
            .withMaxZoom(20);

        // then
        final var error = assertThrows(IllegalArgumentException.class, builder::build);
        assertThat(error)
            .hasMessageContaining("Geometry expression (SRID 4326) must be not null");
    }

    @Test
    void shouldReturnAnErrorWhenWeDoNotSpecifyTheSameGeometryExpressionInBaseQuery() {
        // given
        final var geomExpr = Expressions.path(Geometry.class, "first");
        final var baseQueryGeomExpr = Expressions.path(Geometry.class, "second");
        final var query = baseQuery(baseQueryGeomExpr);
        final var builder = ClusterQueryBuilder.builder(sqlQueryFactory)
            .withBaseQuery(query)
            .withBoundingBox(BOUNDING_BOX)
            .withZoom(10)
            .withMinZoom(5)
            .withMaxZoom(20)
            .withGeometryExpression(geomExpr);

        // then
        final var error = assertThrows(IllegalArgumentException.class, builder::build);
        assertThat(error)
            .hasMessageContaining("The base query should contain the geometry expression");
    }

    @Test
    void shouldReturnAnErrorWhenBaseQueryIsNotProjectable() {
        // given
        final var query = new SQLQuery<>()
            .select(
                new SQLQuery<>()
                    .from(Expressions.path(Geometry.class, "sub_query"))
            )
            .from(Expressions.path(Geometry.class, "base"));
        final var builder = ClusterQueryBuilder.builder(sqlQueryFactory)
            .withBaseQuery(query)
            .withBoundingBox(BOUNDING_BOX)
            .withZoom(10)
            .withMinZoom(5)
            .withMaxZoom(20)
            .withGeometryExpression(geometryExpr);

        // then
        final var error = assertThrows(IllegalArgumentException.class, builder::build);
        assertThat(error)
            .hasMessageContaining("The base query should be a factory expression, such as JavaBean or Constructor projections");
    }

    private static SQLQueryFactory queryFactory() {
        return new SQLQueryFactory(new Configuration(new PostGISTemplateExtension()), () -> null);
    }

    @SuppressWarnings("rawtypes")
    private static SQLQuery<?> baseQuery(Expression<Geometry> geomExpr) {
        return new SQLQuery<>()
            .select(
                Expressions.path(String.class, "name"),
                Expressions.path(String.class, "address"),
                Expressions.path(String.class, "city"),
                Expressions.path(String.class, "postal_code"),
                geomExpr
            )
            .from(Expressions.path(String.class, "places"));
    }

}
