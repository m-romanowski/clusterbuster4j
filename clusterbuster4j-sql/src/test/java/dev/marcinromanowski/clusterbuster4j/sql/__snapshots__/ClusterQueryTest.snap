dev.marcinromanowski.clusterbuster4j.sql.ClusterQueryTest.shouldAllowToCreateClusteringQuery[All values are valid]=with bbox_query as (
  select
    geom as center,
    ? as cluster_size,
    ? as cluster_no,
    name,
    address,
    city,
    postal_code,
    geom
  from
    (
      select
        name,
        address,
        city,
        postal_code,
        geom
      from
        places
    ) as base_query
  where
    ST_Intersects(geom, ST_MakeEnvelope(?, ?, ?, ?, ?))
),
cluster_20 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    bbox_query
),
grouped_cluster_20 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_20
  group by
    cluster
),
cluster_19 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_20
),
grouped_cluster_19 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_19
  group by
    cluster
),
cluster_18 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_19
),
grouped_cluster_18 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_18
  group by
    cluster
),
cluster_17 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_18
),
grouped_cluster_17 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_17
  group by
    cluster
),
cluster_16 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_17
),
grouped_cluster_16 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_16
  group by
    cluster
),
cluster_15 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_16
),
grouped_cluster_15 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_15
  group by
    cluster
),
cluster_14 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_15
),
grouped_cluster_14 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_14
  group by
    cluster
),
cluster_13 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_14
),
grouped_cluster_13 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_13
  group by
    cluster
),
cluster_12 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_13
),
grouped_cluster_12 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_12
  group by
    cluster
),
cluster_11 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_12
),
grouped_cluster_11 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_11
  group by
    cluster
),
cluster_10 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    grouped_cluster_11
),
grouped_cluster_10 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_10
  group by
    cluster
)
select
  cluster_size,
  ST_X(center) as longitude,
  ST_Y(center) as latitude,
  case
    when cluster_size = ? then name
    else null
  end,
  case
    when cluster_size = ? then address
    else null
  end,
  case
    when cluster_size = ? then city
    else null
  end,
  case
    when cluster_size = ? then postal_code
    else null
  end,
  case
    when cluster_size = ? then geom
    else null
  end
from
  grouped_cluster_10


dev.marcinromanowski.clusterbuster4j.sql.ClusterQueryTest.shouldAllowToCreateClusteringQuery[Current zoom should be normalized to min zoom]=with bbox_query as (
  select
    geom as center,
    ? as cluster_size,
    ? as cluster_no,
    name,
    address,
    city,
    postal_code,
    geom
  from
    (
      select
        name,
        address,
        city,
        postal_code,
        geom
      from
        places
    ) as base_query
  where
    ST_Intersects(geom, ST_MakeEnvelope(?, ?, ?, ?, ?))
),
cluster_16 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    bbox_query
),
grouped_cluster_16 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_16
  group by
    cluster
)
select
  cluster_size,
  ST_X(center) as longitude,
  ST_Y(center) as latitude,
  case
    when cluster_size = ? then name
    else null
  end,
  case
    when cluster_size = ? then address
    else null
  end,
  case
    when cluster_size = ? then city
    else null
  end,
  case
    when cluster_size = ? then postal_code
    else null
  end,
  case
    when cluster_size = ? then geom
    else null
  end
from
  grouped_cluster_16


dev.marcinromanowski.clusterbuster4j.sql.ClusterQueryTest.shouldAllowToCreateClusteringQuery[Min, max and current zoom should be normalized to valid values]=with bbox_query as (
  select
    geom as center,
    ? as cluster_size,
    ? as cluster_no,
    name,
    address,
    city,
    postal_code,
    geom
  from
    (
      select
        name,
        address,
        city,
        postal_code,
        geom
      from
        places
    ) as base_query
  where
    ST_Intersects(geom, ST_MakeEnvelope(?, ?, ?, ?, ?))
),
cluster_15 as (
  select
    center,
    cluster_size,
    cluster_no as previous_cluster_no,
    ST_ClusterDBSCAN(center, ?, ?) over () as cluster,
    name,
    address,
    city,
    postal_code,
    geom
  from
    bbox_query
),
grouped_cluster_15 as (
  select
    ST_Centroid(ST_Collect(center)) as center,
    sum(cluster_size) as cluster_size,
    cluster as cluster_no,
    FIRST(name) as name,
    FIRST(address) as address,
    FIRST(city) as city,
    FIRST(postal_code) as postal_code,
    FIRST(geom) as geom
  from
    cluster_15
  group by
    cluster
)
select
  cluster_size,
  ST_X(center) as longitude,
  ST_Y(center) as latitude,
  case
    when cluster_size = ? then name
    else null
  end,
  case
    when cluster_size = ? then address
    else null
  end,
  case
    when cluster_size = ? then city
    else null
  end,
  case
    when cluster_size = ? then postal_code
    else null
  end,
  case
    when cluster_size = ? then geom
    else null
  end
from
  grouped_cluster_15